bindata/assets/alerts/client-go-requests.yaml

```diff
@@ -0,0 +1,53 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  name: client-go
+  namespace: openshift-kube-apiserver
+spec:
+  groups:
+    - name: control-plane-client-go
+      rules:
+        # latency by:
+        # - job (kubelet, apiserver. controller-manager, ...)
+        # - destination host: internal-lb, kubernetes.default service, direct connection ...
+        - record: cluster:rest_client_request_duration_seconds:internal_load_balancer:histogram_quantile
+          labels:
+            quantile: "0.99"
+          expr: |
+            histogram_quantile(0.99,
+              sum(rate(
+                label_replace(rest_client_request_duration_seconds_bucket{verb="GET",url=~"https?://api-int.*"},
+                  "host","$1","url","https?://([^/\\s]+).*")[5m:30s]
+                )
+              ) by (le,host,service,namespace,node)
+            )
+        - record: cluster:rest_client_request_duration_seconds:services:histogram_quantile
+          labels:
+            quantile: "0.99"
+          expr: |
+            histogram_quantile(0.99,
+              sum(rate(
+                label_replace(rest_client_request_duration_seconds_bucket{verb="GET",url!~"https?://(api-int|\\[::1\\]|127\\.0\\.0\\.1|localhost).*"},
+                  "host","$1","url","https?://([^/\\s]+).*")[5m:30s]
+                ) * on (cluster_ip) group_left group(kube_service_info{cluster_ip!~"None",namespace=~"(openshift|kube|default).*"}) by (service,namespace,cluster_ip)
+              ) by (le,host,service,namespace)
+            )
+        - record: cluster:rest_client_request_duration_seconds:localhost:histogram_quantile
+          labels:
+            quantile: "0.99"
+          expr: |
+            histogram_quantile(0.99,
+              sum(rate(
+                label_replace(rest_client_request_duration_seconds_bucket{verb="GET",url=~"https?://(\\[::1\\]|127\\.0\\.0\\.1|localhost).*"},
+                  "host","$1","url","https?://([^/\\s]+).*")[5m:30s]
+                )
+              ) by (le,host,service,namespace)
+            )
+        # error by host and job, the rest_client metrics use the code "<error>" to aggregate
+        # any non-http error and avoid cardinality problems.
+        # xref: https://github.com/kubernetes/kubernetes/blob/66931c9b8f11a3f223ba1890e4f390cff74ce1a6/staging/src/k8s.io/client-go/rest/request.go#L785-L799
+        - record: cluster:rest_client_requests:errors:rate
+          expr: |
+            sum(rate(rest_client_requests_total{code="<error>"}[5m])) by (host,service,namespace)
+            / ignoring (host,service,namespace) group_left
+            sum(rate(rest_client_requests_total{}[5m]))
```

Comment:
![image](https://user-images.githubusercontent.com/6450081/146751713-7db90afd-0e20-4675-9f5b-dd451a06573b.png)

Reply:
I would expect to have 3 different rules for the error rate since we have dedicated rules for latency. Ultimately we want to know the availability of each proxy separately, so we should also differentiate their error rate in telemetry. So I would expect 3 recording rules, one for the internal load balancer, one for the services, and another one for direct connections. Each of them yielding only 1 serie (telemetry requirements), so that we can easily differentiate them in telemetry later and maybe even reuse them to create dedicated alerts with different SLOs.

Reply:
ack